Fix operator precedence in SizesCalcParser.cpp

zakariaridouh opened the pull request:
#### a303c0e7ffc6a5b1441f70db397dd35aa337edb2
<pre>
Fix operator precedence in SizesCalcParser.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=254697">https://bugs.webkit.org/show_bug.cgi?id=254697</a>
&lt;<a href="https://rdar.apple.com/107681323">rdar://107681323</a>&gt;

Reviewed by Tim Nguyen.

Fix operator precedence in `SizesCalcParser.cpp` and write tests for it.
This improves size calculations and could improve experience on
mobile/tablets when rotating from landscape -&gt; horizontal and viceversa.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/e0bcd910832c55445bc6bb4156acd716e0667248">https://chromium.googlesource.com/chromium/src.git/+/e0bcd910832c55445bc6bb4156acd716e0667248</a>

Add in tests to `calc-errors.html` to verify this fix.

* LayoutTests/css3/calc/calc-errors-expected.txt:
* LayoutTests/css3/calc/calc-errors.html:
* Source/WebCore/css/parser/SizesCalcParser.cpp:
(WebCore::SizesCalcParser::handleOperator):

Canonical link: <a href="https://commits.webkit.org/290768@main">https://commits.webkit.org/290768@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7413612f5d7706720606731065a1be413606f3dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96027 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41795 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10938 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18858 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69961 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27484 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93999 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8319 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50287 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8093 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40919 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78393 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98001 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18201 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78969 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18461 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78295 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78166 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/19327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22650 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18208 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23577 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17944 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19729 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on Source/WebCore/css/parser/SizesCalcParser.cpp:
```diff
@@ -62,24 +62,19 @@ static bool operatorPriority(UChar cc, bool& highPriority)
 
 bool SizesCalcParser::handleOperator(Vector<CSSParserToken>& stack, const CSSParserToken& token)
 {
-    // If the token is an operator, o1, then:
```
Can you add `Merge: https://chromium.googlesource.com/chromium/src.git/+/e0bcd910832c55445bc6bb4156acd716e0667248` in commit and also update comments similar to Blink's commit? Thanks for tackling this.

Ahmad-S792 commented on LayoutTests/css3/calc/calc-errors.html:
```diff
@@ -75,8 +75,16 @@
+<div style="width: 100px; width: calc((250px + 40px + 30px + 40px - 60px) / 3);">Operator Precedence Case 3</div>
+<div style="width: 100px; width: calc((500px - 2 / 2 * 40px - 2 * 30px) / 4);">Operator Precedence Case 4</div>
+<div style="width: 100px; width: calc((500px - 2 * 2 / 2 * 40px - 2 * 30px - 60px) / 3);">Operator Precedence Case 5</div>
+<div style="width: 100px; width: calc((500px - 2 * 2 / 2 * 40px - 2 * 30px - 60px) / 3);">Operator Precedence Case 6</div>
```
Isn't Case 5 and 6 same? We should get rid of one and keep just one. @zakariaridouh

zakariaridouh replied:
Good catch.

zakariaridouh committed:
Fix operator precedence in SizesCalcParser.cpp
https://bugs.webkit.org/show_bug.cgi?id=254697
<rdar://107681323>

Reviewed by Tim Nguyen.

Fix operator precedence in `SizesCalcParser.cpp` and write tests for it.
This improves size calculations and could improve experience on
mobile/tablets when rotating from landscape -> horizontal and viceversa.

Merge: https://chromium.googlesource.com/chromium/src.git/+/e0bcd910832c55445bc6bb4156acd716e0667248

Add in tests to `calc-errors.html` to verify this fix.

* LayoutTests/css3/calc/calc-errors-expected.txt:
* LayoutTests/css3/calc/calc-errors.html:
* Source/WebCore/css/parser/SizesCalcParser.cpp:
(WebCore::SizesCalcParser::handleOperator):

Canonical link: https://commits.webkit.org/290768@main
